### PR TITLE
Add tee utilities

### PR DESCRIPTION
**Pull request by kotfic**
This implements a base class which can be subclassed for safe side-effect based actions to take on sys.stdout and sys.stderr.  It is motivated by the desire to replace the JobManager class's [stdout/stderr hijacking](https://github.com/girder/girder_worker/blob/master/girder_worker/utils.py#L100-L105).  This becomes problematic if the task also intends to hijack stdout/stderr (something that happens disturbingly often unfortunately).  As long as all parties are using Tee derived objects then all print() statements will be passed through (including to the original stdout/stderr). 

@zachmullen PTAL

### CODE REVIEW

File: girder_worker_utils/tee.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+def _set_stderr(stream):
+    """Set sys.stderr to a new file-like object.
+
+    This is a private function for re-assigning sys.stderr. It returns
+    the old value of sys.stderr.
+
+    """
+    old = sys.stderr
+    sys.stderr = stream
+    return old
```
Comment (zachmullen) [inline girder_worker_utils/tee.py]: For the above two functions, we could use python's handy swap semantics if we're inclined:

```
old, sys.stderr = sys.stderr, stream
return old
```

File: girder_worker_utils/tee.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        pass
+
+    def __del__(self):
+        # Clean set the stream back to the original on deletion. This
+        # prevents losing the original stream object, and also makes
+        # subclasses very ammenable to contextmanagers.
+        self.set_stream(self._original)
```
Comment (zachmullen) [inline girder_worker_utils/tee.py]: Why is this in `__del__` and not `__exit__`? What's the purpose of having this context manager API?

**Comment by zachmullen** [comment]
Can you show an example of how this is intended to be used downstream?

**Comment by zachmullen** [comment]
I withdraw my comment, I realize these are only intended to be subclassed. The `TeeCapture` is demonstration of its intended use.

**Comment by kotfic** [comment]
tl;dr I am still hashing out exactly how to use this with the JobManager,  i will change this to [WIP]

yeah actually i am not real happy with the __del__ behavior,  this is some bad copy pasta related to figuring out the behavior i want.  Ideally this would work like a context manager when that makes sense,  but also allow for behavior like:

```
foo = TeeCapture()
print('Something')
assert foo.buf == 'Something\n'
```

Maybe it would be better if this kind of behavior was more like this:

```
foo = TeeCapture()
# Do stuff

# explicit reset
foo.reset()
```
There are some concerning issues with that though,  in particular holding onto the Tee object for longer than intended (because of the sys.stdout reference).  That sort of behavior will have to explicitly release the tee which might be problematic.

**Comment by kotfic** [comment]
@zachmullen this is ready for another look.  Hopefully it looks ok - i don't usually have to push code that might fail me on a programming interview question :sweat_smile:

**Comment by kotfic** [comment]
See: https://github.com/girder/girder_worker/pull/260 for an example of how this will be used on the JobManager class

File: girder_worker_utils/tee.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
+import sys
+
+
+def tee_stdout(cls):
```
Comment (zachmullen) [inline girder_worker_utils/tee.py]: Let's document this method and `tee_stderr`, namely to say that it is meant to be used to decorate subclasses of `Tee`.

File: girder_worker_utils/tests/tee_test.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    o1.reset()
+
+    assert sys.stdout == o2
+    assert o2._downstream == original_stdout
+
+
+def test_tee_multiple_tee_objects_reset_o3_then_o1_then_o2():
+    original_stdout = sys.stdout
+    o1, o2, o3 = TeeCapture(), TeeCapture(), TeeCapture()
+
```
Comment (zachmullen) [inline girder_worker_utils/tests/tee_test.py]: Add sanity check:

```
assert sys.stdout == o3
assert o3._downstream == o2
assert o2._downstream == o1
assert o1._downstream == original_stdout
```

Comment (zachmullen) [reply girder_worker_utils/tests/tee_test.py]: Also could you `print('foo')` and make sure each capture gets just one 'foo' and use `capsys` to verify the original stdout also just gets one 'foo'.

Comment (kotfic) [reply girder_worker_utils/tests/tee_test.py]: So as not to muddle the ```test_tee_multiple_tee_objects_reset_o3_then_o1_then_o2``` I've written a separate test that runs earlier to handle this functionality.
